### PR TITLE
drivers: bluetooth: hci: spi: Fix semaphores misplacements

### PR DESCRIPTION
**Pull request by yboutreux**
Fix semaphore misplacement in bt_spi_rx_thread and bt_spi_send functions.

Signed-off-by: Yaël Boutreux <yael.boutreux@st.com>

### CODE REVIEW

**Comment by yboutreux** [comment]
For better understanding, regression coming from here: https://github.com/zephyrproject-rtos/zephyr/commit/7b02e6dc551cba9d30b29b9b6ee5ddc61ae5471d?diff=split
@arnop2 can you please also add @atorgue and @erwango for internal review ?

**Comment by erwango** [comment]
> For context, regression coming from here: zephyrproject-rtos@7b02e6d?diff=split

You mean that configuration board + shield was working fine before ?

**Comment by yboutreux** [comment]
> > For context, regression coming from here: [zephyrproject-rtos/zephyr@7b02e6d](https://github.com/zephyrproject-rtos/zephyr/commit/7b02e6d)?diff=split
> 
> You mean that configuration board + shield was working fine before ?

Can't state that for sure since I'll need to go back in the git history to a state where the MP1 isn't implemented yet. 
This commit it adding the do irq_init_loop.... while{}. This do... while{} is surrounding the existing code, without adding those three lines related to irq_callback and semaphores, which make them dull when going back inside the loops.
It's more like while doing a big step forward, butwith a little step backward which is this tiny mistake.
The term "regression" might not be the most adapted if this was your point, sorry about that :)

**Comment by yboutreux** [comment]
> > > For context, regression coming from here: [zephyrproject-rtos/zephyr@7b02e6d](https://github.com/zephyrproject-rtos/zephyr/commit/7b02e6d)?diff=split
> > 
> > 
> > You mean that configuration board + shield was working fine before ?
> 
> Can't state that for sure since I'll need to go back in the git history to a state where the MP1 isn't implemented yet.
> This commit is adding the do irq_init_loop.... while{}irq_pin_high(). This do... while{} is surrounding the existing code, without adding those three lines related to irq_callback and semaphores, which make them dull when going back inside the loops.
> It's more like while doing a big step forward, but with a little step backward which is this tiny mistake.
> The term "regression" might not be the most adapted if this was your point, sorry about that :)

To be more clear, those specific points (both the two semaphores and IRQ_callback management) were well implemented before, but I don't speak about the rest of the content of the commit

**Comment by erwango** [comment]
ok, it could actually have been a difference between disco board and shield that we didn't take into account. Thanks for the clarification

File: drivers/bluetooth/hci/spi.c [hunk 1]
```diff
@@ -309,14 +309,14 @@ static void bt_spi_rx_thread(void)
...
-		gpio_pin_disable_callback(irq_dev, GPIO_IRQ_PIN);
-		k_sem_take(&sem_busy, K_FOREVER);
 
 		BT_DBG("");
 
 		do {
+			k_sem_take(&sem_request, K_FOREVER);
+			/* Disable IRQ pin callback to avoid spurious IRQs */
+			gpio_pin_disable_callback(irq_dev, GPIO_IRQ_PIN);
+			k_sem_take(&sem_busy, K_FOREVER);
```
Comment (arnop2) [inline drivers/bluetooth/hci/spi.c]: Seems a workaround for something not well understood. 
IRQ pin is used to wake up master for reception right? initial code seem waiting the release of the GPIO before going in sleep ( blocked on sem_request) waiting next IRQ.  
Look like the IRQ on the STM32MP1 is on level not on rising edge...
To discuss to clarify  behavior...

Comment (yboutreux) [reply drivers/bluetooth/hci/spi.c]: IRQ pin is set on level low triggering, as you you guessed. Looking into it

Comment (erwango) [reply drivers/bluetooth/hci/spi.c]: Ok, this would explain the difference of behavior on MP1 and the fact this issue was not seen on L4. Thanks @arnop2

Comment (yboutreux) [reply drivers/bluetooth/hci/spi.c]: @erwango I didn't specified but the L4 has the same definition trough board DTS 
```
spbtle-rf@0 {
		compatible = "zephyr,bt-hci-spi";
		reg = <0>;
		reset-gpios = <&gpioa 8 0>;
		irq-gpios = <&gpioe 6 **0**>; //That means GPIO_DIR_IN + GPIO_INT_LEVEL + GPIO_INT_ACTIVE_LOW
		spi-max-frequency = <2000000>;
		label = "SPBTLE-RF";
	};
```
So this doesn't explain the difference between MP1 and L4

Comment (yboutreux) [reply drivers/bluetooth/hci/spi.c]: EDIT : IRQ is defined low level active in DTS, but it seems this is override in hci's spi.c code 
`#if defined(CONFIG_BT_SPI_BLUENRG)
			   GPIO_PUD_PULL_DOWN |
#endif
			   GPIO_DIR_IN | GPIO_INT |
			   GPIO_INT_EDGE | GPIO_INT_ACTIVE_HIGH);`

**Comment by yboutreux** [comment]
Updated commit, removed do{}while(irq_pin_high()) loop which is causing wrong behaviors with BLUENRG shield X-NUCLEO-IDB04A1

**Comment by erwango** [comment]
> Updated commit, removed do{}while(irq_pin_high()) loop which is causing wrong behaviors with BLUENRG shield X-NUCLEO-IDB04A1

Can you test with disco board?

**Comment by yboutreux** [comment]
> > Updated commit, removed do{}while(irq_pin_high()) loop which is causing wrong behaviors with BLUENRG shield X-NUCLEO-IDB04A1
> 
> Can you test with disco board?

From what I tested (will be : [before edit] | [after edits]) :

- Beacon : Both OK before and after edits

- Central : both stops when the L4 find a signal stronger than -70dBm (seems like it can't connect successfully, or maybe I don't a compatible peripheral to which L4 can connect to)

-Peripheral : Both crashes when I try to connect with and Android phone + nRF Connect App, either not working or waiting for an other kind of device connection. But without the loop-part removal, some "<wrn> bt_att: Ignoring unexpected request" are printed on the console 

-ipsp : Both working until the waiting for connection part, didn't try further the connection part

-scan_adv : Both "Advertising failed to start"

**Comment by yboutreux** [comment]
EDIT: those commits don't solve bluetooth init problems with the STM32MP1, but correct mistakes about semaphore placement made during merges.

**Comment by erwango** [comment]
Seems correct indeed.
But wait let us get back on a working status on the driver (currently impacted by a non regression).
Then we'll see how this change impacts.

**Pull request closed without merging**